File: extension/src/link/queue.ts
import { IMPORT_MAX_AGE_S, IMPORT_MAX_FUTURE_S, QUEUE_CAPACITY } from "./constants";

export type QueuedPhase = "work" | "short" | "long";

export interface QueuedSession {
  client_id: string;
  type: QueuedPhase;
  duration: number;
  completed: true;
  start?: number;
  tag?: string;
}

const PHASES = new Set<string>(["work", "short", "long"]);

function asPhase(value: unknown): QueuedPhase | null {
  return typeof value === "string" && PHASES.has(value) ? (value as QueuedPhase) : null;
}

export function parseQueuedSession(row: unknown): QueuedSession | null {
  if (typeof row !== "object" || row === null) return null;
  const record = row as Record<string, unknown>;
  const clientId = String(record.client_id ?? "");
  const type = asPhase(record.type);
  const duration = Number(record.duration);
  if (clientId.length === 0 || type === null || !Number.isInteger(duration) || duration <= 0) return null;
  const item: QueuedSession = { client_id: clientId, type, duration, completed: true };
  if (record.start !== undefined && record.start !== null) {
    const start = Number(record.start);
    if (Number.isInteger(start) && start > 0) item.start = start;
  }
  const tag = String(record.tag ?? "");
  if (tag.length > 0) item.tag = tag;
  return item;
}

export class SessionQueue {
  items: QueuedSession[] = [];

  constructor(
    private readonly capacity = QUEUE_CAPACITY,
    rows: unknown[] = [],
  ) {
    this.load(rows);
  }

  load(rows: unknown[]): void {
    this.items = [];
    for (const row of rows) {
      if (this.items.length >= this.capacity) break;
      const item = parseQueuedSession(row);
      if (item !== null) this.items.push(item);
    }
  }

  count(): number {
    return this.items.length;
  }

  empty(): boolean {
    return this.items.length === 0;
  }

  at(index: number): QueuedSession | undefined {
    return this.items[index];
  }

  toRows(): QueuedSession[] {
    return this.items.map((item) => {
      const row: QueuedSession = {
        client_id: item.client_id,
        type: item.type,
        duration: item.duration,
        completed: true,
      };
      if (item.start !== undefined && item.start > 0) row.start = item.start;
      if (item.tag !== undefined && item.tag.length > 0) row.tag = item.tag;
      return row;
    });
  }

  enqueue(clientId: string, type: QueuedPhase, durationSec: number, startEpoch: number | null, tag = ""): boolean {
    if (clientId.length === 0) return false;
    if (asPhase(type) === null) return false;
    const duration = Math.floor(durationSec);
    if (!Number.isInteger(duration) || duration <= 0) return false;
    if (this.items.length >= this.capacity) this.dropOldest();
    const item: QueuedSession = { client_id: clientId, type, duration, completed: true };
    if (startEpoch !== null && Number.isInteger(startEpoch) && startEpoch >= 0) item.start = startEpoch;
    if (tag.length > 0) item.tag = tag;
    this.items.push(item);
    return true;
  }

  dropOldest(): void {
    if (this.items.length > 0) this.items.shift();
  }

  dropByClientId(clientIds: string[]): number {
    if (clientIds.length === 0 || this.items.length === 0) return 0;
    const drop = new Set(clientIds);
    const kept: QueuedSession[] = [];
    let dropped = 0;
    for (const item of this.items) {
      if (drop.has(item.client_id)) {
        dropped += 1;
        continue;
      }
      kept.push(item);
    }
    this.items = kept;
    return dropped;
  }

  stripImplausibleStarts(nowEpoch: number): number {
    if (!Number.isInteger(nowEpoch) || nowEpoch <= 0 || this.items.length === 0) return 0;
    const maxFuture = nowEpoch + IMPORT_MAX_FUTURE_S;
    const minStart = nowEpoch - IMPORT_MAX_AGE_S;
    let stripped = 0;
    for (const item of this.items) {
      if (item.start === undefined) continue;
      if (item.start > maxFuture || item.start < minStart) {
        delete item.start;
        stripped += 1;
      }
    }
    return stripped;
  }
}

File: omarchy-plugin/lib/pomo_link/queue.py
"""Bounded offline session history for POST /api/sessions/import."""

from __future__ import annotations

from .constants import IMPORT_MAX_AGE_S, IMPORT_MAX_FUTURE_S, PHASES, QUEUE_CAPACITY
from .persist import atomic_write, load_json


def _valid_type(value):
    return value in PHASES


class SessionQueue:
    def __init__(self, path, capacity=QUEUE_CAPACITY):
        self.path = path
        self.capacity = capacity
        self.items = []
        self.load()

    def count(self):
        return len(self.items)

    def empty(self):
        return not self.items

    def at(self, index):
        return self.items[index]

    def load(self):
        self.items = []
        data = load_json(self.path)
        if not isinstance(data, dict):
            return
        rows = data.get("sessions")
        if not isinstance(rows, list):
            return
        skipped = 0
        for row in rows:
            if not isinstance(row, dict):
                skipped += 1
                continue
            if len(self.items) >= self.capacity:
                skipped += 1
                continue
            client_id = str(row.get("client_id") or "")
            typ = str(row.get("type") or "")
            duration = int(row.get("duration") or 0)
            if not client_id or not _valid_type(typ) or duration <= 0:
                skipped += 1
                continue
            item = {
                "client_id": client_id,
                "type": typ,
                "duration": duration,
                "completed": True,
            }
            if "start" in row and row["start"] is not None:
                try:
                    start = int(row["start"])
                except (TypeError, ValueError):
                    start = 0
                if start > 0:
                    item["start"] = start
            tag = str(row.get("tag") or "")
            if tag:
                item["tag"] = tag
            self.items.append(item)
        _ = skipped

    def save(self):
        rows = []
        for item in self.items:
            row = {
                "client_id": item["client_id"],
                "type": item["type"],
                "duration": int(item["duration"]),
                "completed": True,
            }
            if item.get("start"):
                row["start"] = int(item["start"])
            if item.get("tag"):
                row["tag"] = item["tag"]
            rows.append(row)
        atomic_write(self.path, {"sessions": rows}, mode=0o600)
        return True

    def enqueue(self, client_id, typ, duration_sec, start_epoch, tag=""):
        if not client_id:
            return False
        if not _valid_type(typ):
            return False
        duration_sec = int(duration_sec)
        if duration_sec <= 0:
            return False
        if len(self.items) >= self.capacity:
            self.drop_oldest()
        item = {
            "client_id": str(client_id),
            "type": typ,
            "duration": duration_sec,
            "completed": True,
        }
        if start_epoch is not None and int(start_epoch) >= 0:
            item["start"] = int(start_epoch)
        if tag:
            item["tag"] = str(tag)
        self.items.append(item)
        self.save()
        return True

    def drop_oldest(self):
        if self.items:
            self.items.pop(0)

    def drop_by_client_id(self, client_ids):
        if not client_ids or not self.items:
            return 0
        drop = set(client_ids)
        kept = []
        dropped = 0
        for item in self.items:
            if item.get("client_id") in drop:
                dropped += 1
                continue
            kept.append(item)
        self.items = kept
        if dropped:
            self.save()
        return dropped

    def strip_implausible_starts(self, now_epoch):
        now_epoch = int(now_epoch or 0)
        if now_epoch <= 0 or not self.items:
            return 0
        max_future = now_epoch + IMPORT_MAX_FUTURE_S
        min_start = now_epoch - IMPORT_MAX_AGE_S
        stripped = 0
        for item in self.items:
            start = item.get("start")
            if start is None:
                continue
            if start > max_future or start < min_start:
                item.pop("start", None)
                stripped += 1
        if stripped:
            self.save()
        return stripped

    def clear(self):
        self.items = []
        self.save()
